src/dbos-executor.ts

```diff
@@ -852,6 +853,25 @@ export class DBOSExecutor implements DBOSExecutorContext {
       }
     }
 
+    const handleWorkflowError = async (err: Error) => {
+      // Record the error.
+      const e = err as Error & { dbos_already_logged?: boolean };
+      this.logger.error(e);
+      e.dbos_already_logged = true;
+      if (wCtxt.isTempWorkflow) {
+        internalStatus.workflowName = `${DBOSExecutor.tempWorkflowName}-${wCtxt.tempWfOperationType}-${wCtxt.tempWfOperationName}`;
+      }
+      internalStatus.error = DBOSJSON.stringify(serializeError(e));
+      internalStatus.status = StatusString.ERROR;
+      if (internalStatus.queueName && !this.isDebugging) {
+        await this.systemDatabase.dequeueWorkflow(workflowID, this.#getQueueByName(internalStatus.queueName));
+      }
+      if (!this.isDebugging) {
+        await this.systemDatabase.recordWorkflowError(workflowID, internalStatus);
+      }
```

Comment:
Note that we should put `dequeueWorkflow` and `recordWorkflowError` in one transaction. We can solve it in a separate PR.